libyul/AST.h

```diff
@@ -36,38 +36,69 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct DebugData
+class DebugData
 {
+public:
+	typedef typename std::shared_ptr<DebugData const> ConstPtr;
+
 	explicit DebugData(
-		langutil::SourceLocation _nativeLocation,
+		langutil::SourceLocation _nativeLocation = {},
 		langutil::SourceLocation _originLocation = {},
 		std::optional<int64_t> _astID = {}
 	):
-		nativeLocation(std::move(_nativeLocation)),
-		originLocation(std::move(_originLocation)),
-		astID(std::move(_astID))
+		m_nativeLocation(std::move(_nativeLocation)),
+		m_originLocation(std::move(_originLocation)),
+		m_astID(_astID)
 	{}
 
-	static std::shared_ptr<DebugData const> create(
-		langutil::SourceLocation _nativeLocation = {},
-		langutil::SourceLocation _originLocation = {},
-		std::optional<int64_t> _astID = {}
-	)
+	DebugData& nativeLocation(langutil::SourceLocation _nativeLocation)
+	{
+		m_nativeLocation = _nativeLocation;
+		return *this;
+	}
+	[[nodiscard]] langutil::SourceLocation const& nativeLocation() const { return m_nativeLocation; }
+	[[nodiscard]] langutil::SourceLocation& nativeLocation() { return m_nativeLocation; }
+
+	DebugData& originLocation(langutil::SourceLocation _originLocation)
 	{
+		m_originLocation = _originLocation;
+		return *this;
+	}
+	[[nodiscard]] langutil::SourceLocation const& originLocation() const { return m_originLocation; }
+	[[nodiscard]] langutil::SourceLocation& originLocation() { return m_originLocation; }
+
+	DebugData& astID(int64_t _astID)
+	{
+		m_astID = _astID;
+		return *this;
+	}
+	[[nodiscard]] std::optional<int64_t> const& astID() const { return m_astID; }
+	[[nodiscard]] std::optional<int64_t>& astID() { return m_astID; }
+
+	[[nodiscard]] DebugData::ConstPtr constPtr()
+	{
+		static DebugData::ConstPtr emptyDebugData{std::make_shared<DebugData const>(DebugData())};
+		if (m_nativeLocation == emptyDebugData->m_nativeLocation &&
+			m_originLocation == emptyDebugData->m_originLocation &&
+			m_astID == emptyDebugData->m_astID)
+			return emptyDebugData;
+
```

Comment:
Maybe we could add here a simple shared-ptr reuse mechanism (similar to the `emptyDebugData` stuff). This mechanism could be used to will always get back a unique shared-ptr pointing to the same data.

Reply:
Why? Is it likely that we'll create the same data all over again? You're trying to optimize things before the usage patterns are even clear :-). Just keep it at the plain ``create`` mechanism we had before - we can still adjust that, if it becomes an issue.
But I'd say it's pretty unlikely that we create the exact same debug data multiple times in the first place.